### DISCUSSION TITLE
Explicit network structure with Virus on a Network model

### DISCUSSION THREAD

**Discussion by clarkpetri** [mesa/mesa#1427] [Q&A] [answered]
The original Virus on a Network example runs fine. As I tend to do, I've stripped many features from it to simplify the model before I build new features on it for my purposes. However, I want to feed it an explicit network structure and not generate a randomly rewired Erdos Renyi graph. The code screenshot below includes many of the simplifications I made and you'll see the commented in/out attempts. The Mesa visualizer does produce the E-R graph and though it's unexciting due to limited features, if I start the model it'll run. However, the later attempts on lines 56 or 58/59 don't visualize and the model won't step, yet it provides no errors in the terminal. Thoughts? As always, many thanks to this wonderful community.

<img width="711" alt="Screen Shot 2022-09-29 at 4 57 17 PM" src="https://user-images.githubusercontent.com/99350552/193140602-fc52ab93-3827-4a87-b577-f5ade6d51a2f.png">

**Comment by tpike3** [comment]
That should work, but you don't need `self.G = net`.

To debug I would run a print statement after you add the edges something like `print(self.G.nodes, self.G.edges)` to see what you are actually loading in `'self.G` and compare that to what is working.

**Reply by rht** [reply]
`self.G = net` for a custom graph should work. The point is that @clarkpetri wants to use a network initialization that is different from Erdos-Renyi graph. I can't help debug because to do that I have to re-type everything in the screenshot.

**Reply by clarkpetri** [reply]
Here's a non-screenshot. Additionally, I added the recommended print statement and regardless of network construction, the print statement works and shows [1, 2, 3] [(1, 2), (1, 3), (2, 3)]. Still, the model only works with Erdos Renyi. 

Thanks. I'm scraping large amounts of Twitter data with an academic API and looking to construct specific networks with that data, hence my interest in feed Mesa a specific network.

`class VirusOnNetwork(Model):
    """A virus model with some number of agents"""

    def __init__(
        self,
        initial_outbreak_size=1,
        virus_spread_chance=0.4,
        virus_check_frequency=0.4,
        recovery_chance=0.3,
        gain_resistance_chance=0.5,
    ):

        net = nx.Graph()
        edges = [(1,2),(2,3),(3, 1)]
        net.add_edges_from(edges)

        #self.num_nodes = num_nodes
        prob = 1

        # THIS WORKS
        # self.G = nx.erdos_renyi_graph(3, p=prob)
        
        # THESE DO NOT
        self.G = net 

        #self.G = nx.Graph()
        #self.G.add_edges_from(edges)

        print(self.G.nodes, self.G.edges)

        self.grid = NetworkGrid(self.G)
        self.schedule = RandomActivation(self)
        #self.initial_outbreak_size = (
        #    initial_outbreak_size if initial_outbreak_size <= num_nodes else num_nodes
        #)
        self.virus_spread_chance = virus_spread_chance
        self.virus_check_frequency = virus_check_frequency
        self.recovery_chance = recovery_chance
        self.gain_resistance_chance = gain_resistance_chance

        self.datacollector = DataCollector(
            {
                "Infected": number_infected,
                "Susceptible": number_susceptible,
                "Resistant": number_resistant,
            }
        )`

**Comment by rht** [comment]
You forgot to do this:
```python
net.add_nodes_from(range(self.num_nodes))
```
There is actually an error message from the console of the web browser, which says that the node doesn't exist.

**Reply by rht** [reply]
Huh, nvm, the error I encountered is probably specific to my code, since the `num_nodes` is 10, but you have removed this argument in your model. Anyway, you can check your browser console for the error message.

**Comment by rht** [comment, marked as answer]
Try `edges = [(0,1),(1,2),(2, 0)]` instead of `edges = [(1,2),(2,3),(3, 1)]`.

**Reply by clarkpetri** [reply]
As always, thank you! Works and I manually built out more. I hadn't considered zero indexing being the issue, especially since I had no errors at the terminal. 

Do you have any insight as to what causes this? With NetworkX facilitates a variety of node naming conventions, yet if I add `edges = [(0,1),(1,2),(2, 0),(2,'a')]` the network won't display again.

**Reply by rht** [reply]
It could be part of the Mesa code assumes the 0 indexing. It shouldn't make this assumption, and is a bug. You can raise an issue and I will investigate.
